cpg-analysis/src/main/kotlin/de/fraunhofer/aisec/cpg/query/FlowQueries.kt

```diff
@@ -0,0 +1,654 @@
+    findAllPossiblePaths: Boolean = true,
+    direction: AnalysisDirection = Forward(),
+    vararg sensitivities: AnalysisSensitivity = FieldSensitive() + ContextSensitive(),
+    scope: AnalysisScope = Interprocedural(),
+    earlyTermination: (Node, Context) -> Boolean = { _, _ -> false },
+    predicate: (Node) -> Boolean,
+): FulfilledAndFailedPaths {
+
+    return this.followXUntilHit(
+        x = { currentNode, ctx, path ->
+            if (direction is Forward) {
+                    currentNode.nextDFGEdges
+                } else {
+                    currentNode.prevDFGEdges
+                }
+                .filter { edge ->
+                    // TODO: There's a problem with the context which should be copied for each
+                    // (potential) path.
+                    scope.followEdge(currentNode, edge, ctx) &&
+                        sensitivities.all { it.followEdge(currentNode, edge, ctx) }
+                }
+                .map { direction.unwrapNextStepFromEdge(it) }
+        },
+        collectFailedPaths = collectFailedPaths,
+        findAllPossiblePaths = findAllPossiblePaths,
+        earlyTermination = earlyTermination,
+        predicate = predicate,
+    )
+}
+
+/**
+ * Follows the [de.fraunhofer.aisec.cpg.graph.edges.flows.Dataflow] edges from [startNode] in the
+ * given [direction] (default: Forward analysis) until reaching a node fulfilling [predicate].
+ *
+ * The interpretation of the analysis result can be configured as must analysis (all paths have to
+ * fulfill the criterion) or may analysis (at least one path has to fulfill the criterion) by
+ * setting the [type] parameter (default: [AnalysisType.MAY]). Note that this only reasons about
+ * existing DFG paths, and it might not be sufficient if you actually want a guarantee that some
+ * action always happens with the data. In this case, you may need to check the [executionPath].
+ *
+ * The [sensitivities] can also be configured to a certain extent. The analysis can be run as
+ * interprocedural or intraprocedural analysis. If [earlyTermination] is not `null`, this can be
+ * used as a criterion to make the query fail if this predicate is fulfilled before [predicate].
+ */
+fun dataFlow(
+    startNode: Node,
+    direction: AnalysisDirection = Forward(),
+    type: AnalysisType = AnalysisType.MAY,
+    vararg sensitivities: AnalysisSensitivity = FieldSensitive() + ContextSensitive(),
+    scope: AnalysisScope = Interprocedural(),
+    verbose: Boolean = true,
+    earlyTermination: ((Node) -> Boolean)? = null,
+    predicate: (Node) -> Boolean,
+): QueryTree<Boolean> {
+    val collectFailedPaths = type == AnalysisType.MUST || verbose
+    val findAllPossiblePaths = type == AnalysisType.MUST || verbose
+    val earlyTermination = { n: Node, ctx: Context -> earlyTermination?.let { it(n) } == true }
+
+    val evalRes =
+        if (direction is Bidirectional) {
+                arrayOf(Forward(), Backward())
```

Comment:
There must be a better way to achieve this (and ideally get rid of the `fold`). Any smart ideas?